### PR TITLE
Run2-TB42 Add a new version of TB (October 19 v1)

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

These codes are for test beam with HGCal prototype

#### PR validation:

Tried with some private runs with dedicated scenario and configuration

#### if this PR is a backport please specify the original PR:

This code is meant for master with no intention of back porting

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.cc [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+#include "DetectorDescription/Core/interface/DDCurrentNamespace.h"
+#include "DetectorDescription/Core/interface/DDSplit.h"
+#include "Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h"
+#include "CLHEP/Units/GlobalPhysicalConstants.h"
+#include "CLHEP/Units/GlobalSystemOfUnits.h"
+
+//#define EDM_ML_DEBUG
+
+DDHGCalTBModuleX::DDHGCalTBModuleX() : factor_(0.5*sqrt(2.0)),
+				       tan30deg_(tan(30.0*CLHEP::deg)) {
```
Comment (cvuosalo) [inline Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.cc]: Other pull requests are replacing CLHEP. Why not this one?
Also, `0.5*sqrt(2.0)` should probably be a constant:
`constexpr double sqr2over2 = 0.5*sqrt(2.0);`

File: Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.cc [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+#endif
+    } else if (layerSense_[ly] > 0) {
+      positionSensitive(zz, copy, layerSense_[ly], rMax_, maxModule_[ly],
+			ignoreCenter, module, cpv);
+    }
+    ++copyNumber_[ii];
+    zi     = zo;
+  } // End of loop over layers in a block
+
+  if (fabs(thickTot-totalWidth) < 0.00001) {
```
Comment (cvuosalo) [inline Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.cc]: Need to use a constant here:
`constexpr double tolerance = 0.00001;`
Maybe it should have a more specific name.

File: Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.cc (hunk 1 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.cc]: Done

File: Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.cc (hunk 2 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.cc]: Done

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+				 const DDLogicalPart&, DDCompactView&);
+  void          positionSensitive(double zpos, int copyIn, int type, 
+				  double rmax, int ncrMax, bool ignoreCenter,
+				  const DDLogicalPart&, DDCompactView& cpv);
+
+private:
+
+  static constexpr double factor_ = 0.5*sqrt(2.0);
+  static constexpr double tolerance_ = 0.00001;
+  const double            tan30deg_;
```
Comment (cvuosalo) [inline Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h]:
```suggestion
  static constexpr double tan30deg_ = tan(30._deg);
```

Comment (cvuosalo) [reply Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h]: Isn't `tan30deg_` a fixed constant like `factor`? Shouldn't it be `static`? You might have to include <cmath> in this header file for the tan function. Is that a problem?

Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h]: Moved

**Comment by bsunanda** [comment]
@cvuosalo clang gives an error message for the factor_. sort is not a constexpr. This states:
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-03-18-1100/src/Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h:40:27: error: constexpr variable 'factor_' must be initialized by a constant expression
  static constexpr double factor_    = 0.5*sqrt(2.0);
                          ^            ~~~~~~~~~~~~~
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-03-18-1100/src/Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h:40:44: note: non-constexpr function 'sqrt' cannot be used in a constant expression
  static constexpr double factor_    = 0.5*sqrt(2.0);

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
+#ifndef HGCalCommonData_DDHGCalTBModuleX_h
+#define HGCalCommonData_DDHGCalTBModuleX_h
+
+#include <map>
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h]: this include is not needed

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 5]
```diff
@@ -70,10 +70,20 @@ void HGCalGeomParameters::loadGeometryHexagon(const DDFilteredView& _fv,
...
       if (itr == layers.end()) {
-        const DDTubs & tube = static_cast<DDTubs>(sol);
-        double rin = HGCalParameters::k_ScaleFromDDD*tube.rIn();
-        double rout= HGCalParameters::k_ScaleFromDDD*tube.rOut();
-        double zp  = HGCalParameters::k_ScaleFromDDD*fv.translation().Z();
+	double rin(0), rout(0);
+	if ((sol.shape() == DDSolidShape::ddpolyhedra_rz) ||
+	    (sol.shape() == DDSolidShape::ddpolyhedra_rrz)) {
+	  const DDPolyhedra& polyhedra = static_cast<DDPolyhedra>(sol);
+	  std::vector<double> rmin = polyhedra.rMinVec();
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: `const std::vector<double>& rmin` to avoid an unnecessary copy (also for the next one)

Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: done

File: Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h (hunk 4 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/DDHGCalTBModuleX.h]: Done

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
+upgrade

**Comment by cvuosalo** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
@bsunanda not sure about which previous use has been of this geometry, now it is modified. Technically the PR is ok for me to go

**Comment by bsunanda** [comment]
@fabiocos The earlier version was approximate and this one takes care of the absorber structure correctly. The old version need to go. There are 2 more setups and we are working on these 2 to appear soon.

**Comment by fabiocos** [comment]
+1

**Pull request merged**